File: Baekjoon_Solve/Recursion/5904.java
// Moo 게임 (5904번)
/*
  문제: https://www.acmicpc.net/problem/5904
  재귀, 분할 정복
  N이 S(k) 중에서 몇 번째인지 찾고, 다시 N을 규칙에 따라서 분할했다.
*/


import java.io.*;

public class Main {

    static void recur(int n, int k) {
        if (n >= N) {
            M = (n - k) / 2;
            K = k;
            return;
        }
        recur(n + k + 1 + n, k + 1);
    }

    static void recur2(int k, int n, int m) {
        if (1 <= n - m && n - m <= k) {
            if (n - m == 1) {
                ans = "m";
            } else {
                ans = "o";
            }
        } else if (n <= m) {
            recur2(k - 1, n, (m - (k - 1)) / 2);
        } else {
            recur2(k - 1, n - m - k, (m - (k - 1)) / 2);
        }
    }

    static int N, M, K;
    static String ans;
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        N = Integer.parseInt(br.readLine());
        recur(3, 3);
        recur2(K, N, M);
        System.out.println(ans);
    }
}

File: Baekjoon_Solve/Recursion/5904/5904.py
import sys
input = sys.stdin.readline

N = int(input().rstrip())
K = 0
M = 0


# N이 몇 단계 moo 수열인지 찾는 함수
# n: moo 수열의 길이(S(k)), k: 단계, m: S(k - 1)의 길이
def recur(n: int, k: int) -> None:
    global K, M
    if n >= N:
        K = k
        M = (n - (k + 3)) >> 1
        return
    recur(2 * n + k + 4, k + 1)


# 재귀 호출을 통해 N번째 글자를 찾는 함수
# k 단계를 0 단계로 낮춰가면서 호출함
def recur2(n: int, m: int, k: int) -> None:
    global ans
    if k == 0:
        ans = "moo"[n - 1]
        return
    if m < n < m + k + 3:
        if n == m + 1:
            ans = "m"
        else:
            ans = "o"
    elif n <= m:
        recur2(n, (m - (k + 2)) // 2, k - 1)
    else:
        recur2(n - m - (k + 3), (m - (k + 2)) // 2, k - 1)


ans = ""
recur(3, 0)
recur2(N, M, K)
print(ans)